### ISSUE TITLE
More elegant `select` API

### ISSUE THREAD

**Issue by flybayer** [prisma/prisma#3796]
## Problem

Currently in Blitz apps I'm manually selecting all the fields I need to prevent unwanted fields from being exposed to the API.

The problem is that this is super cumbersome to write as an object with `: true` for every field.

```ts
  const order = await prisma.order.findOne({
    where,
    select: {
      id: true,
      createdAt: true,
      updatedAt: true,
      chargeStatus: true,
      amount: true,
      customer: {
        select: {
          id: true,
          firstName: true,
          lastName: true,
          email: true,
          phone: true,
        },
      },
      orderGroup: {
        select: {
          name: true,
        },
      },
    },
  })
```

## Suggested solution

Something easier, perhaps accepting an array of strings.

```ts
  const order = await prisma.order.findOne({
    where,
    select: {
      fields: ['id', 'createdAt', 'updatedAt', 'chargeStatus', 'amount'],
      customer: {
        select: {
          fields: ['id', 'firstName', 'lastName', 'email', 'phone']
        }
      },
      orderGroup: {
        select: {
          fields: ['name']
        }
      },
    },
  })
```

And it would be sweet to have a shortcut without `fields`:

```ts
  const order = await prisma.order.findOne({
    where,
    select: {
      fields: ['id', 'createdAt', 'updatedAt', 'chargeStatus', 'amount'],
      customer: {
        select: ['id', 'firstName', 'lastName', 'email', 'phone']
      },
      orderGroup: {
        select: ['name']
      },
    },
  })
```

## Alternatives

I haven't thought of anything else.

## Additional context
-

**Comment by flybayer** [comment]
Another way to solve the root problem is a way to hide fields in query results by default: https://github.com/prisma/prisma-client-js/issues/649#issuecomment-700776231

That would reduce the current pain experienced from this issue, but of course we'll always still have a need for `select`.

**Comment by Sytten** [comment]
My take on it is that is a vastly different syntax and it will cause beginners to be even more lost when they start (plus it confuses the autocomplete). If this is considered seriously, I would do the following:
```typescript
const order = await prisma.order.findOne({
    where,
    select: {
      FIELDS: ['id', 'createdAt', 'updatedAt', 'chargeStatus', 'amount'],
      customer: {
        select: {
          FIELDS: ['id', 'firstName', 'lastName', 'email', 'phone']
        }
      },
      orderGroup: {
        select: {
          FIELDS: ['name']
        }
      },
    },
  })
```
Similar to `AND`, `OR`, etc in the `where` clause. It makes it clear that this is a Prisma "advanced" thing and it makes sure we don't have a naming conflict on models that might have a column named `fields`.

**Comment by thebiglabasky** [comment]
It looks like [this](https://github.com/prisma/prisma-client-js/issues/649) would solve that problem to a degree. Is that a fair assumption?
I otherwise don't see this coming through anytime soon, as it would create confusion across multiple ways of querying the same thing.

**Comment by flybayer** [comment]
Yes, that solves the problem to a degree, but regardless the current select API is super verbose and gets annoying very quickly.

I wouldn't say this is high priority, but I'd definitely love to have this addressed at some point.

**Comment by thebiglabasky** [comment]
The problem is that it's either a breaking change, or providing alternative ways to do the same thing which makes it unlikely to change unless we were to revisit the client API significantly (like a kind of v3 or so). That's why I'd lean on another solution that mitigates the pain.
I believe we should push the `exclude` solution to see how big the pain stays and reconsider later in this context. @matthewmueller do you have additional thoughts there?

**Comment by nikolasburk** [comment]
For anyone finding this issue, I just quickly want to drop in and mention that a user on [Slack](https://prisma.slack.com/archives/CA491RJH0/p1612853767386400?thread_ts=1612836270.383400&cid=CA491RJH0) today posted a nice workaround pattern for an `exclude` API:

![image](https://user-images.githubusercontent.com/4058327/107352819-97f1a200-6acc-11eb-95bd-4ac5e0054030.png)

We'll also [document](https://github.com/prisma/docs/issues/1280) this pattern soon.

**Comment by un33k** [comment]
@flybayer I am not sure if using fields as string will disable intelisense or not as per your recommendation above.  How about something like the following recommendation I've made on a different ticket ?

If we don't allow select with mixed `true & false`, then we are limited to only a small set of `true` or `false`.  As we only want to exclude a few fields such as `password` or include a few fields such as `id`, otherwise, we want everything.  If so, it will not be too cumbersome at all. 

What do you think ?  
https://github.com/prisma/prisma/issues/5042#issuecomment-855440699

**Comment by henk23** [comment]
A bit sad to see that 4 years and 3 major versions later, this is still an open issue.

I've never worked with Prisma, was looking for a new ORM to try, read through the docs and stumbled upon the convoluted select syntax, then found this issue looking for a better way to do it...

Since "selecting stuff" is a big part of what I plan to do with an ORM, I like the code around it to be very concise, so this is kind of a deal breaker for me, and looking at the thumbs-up count of the OP, maybe it's time to give this a higher priority?

If anyone comes here looking for a way to exclude fields, this has been implemented and documented here: https://www.prisma.io/docs/orm/prisma-client/queries/excluding-fields

Having worked with directus, I got used to beautifully selecting fields like this:

```js
fields: ['*'] // Select all top-level fields
fields: ['*.*'] // Select all top-level fields and all second-level relational fields
fields: ['id', 'customer.email'] // Select order id and the email field of the customer relation
fields: ['id', 'amount', 'customer.*'] // Select order id and amount and all fields of the customer relation
```

Look how they all fit in just one line of code 🥰 Isn't this super desirable? ^^

For the time being, I will probably help myself out with a helper function like this. Feel free to copy&paste. :) This sadly does not handle placeholder asterisk * magic, but it's a start.

```js
function prismaSelectify(keys) {
  return keys.reduce((acc, key) => {
    const parts = key.split('.');

    let current = acc;

    parts.forEach((part, index) => {
      // If it's the last part of the key, assign it to true or nested object
      if (index === parts.length - 1) {
        current[part] = true;
      } else {
        // If it isn't already an object, initialize it as one with a "select" key
        current[part] = current[part] || { select: {} };
        current = current[part].select;
      }
    });

    return acc;
  }, {});
}

prismaSelectify([['a', 'b.x', 'c']])

// returns
{
  a: true,
  b: {
    select: {
      x: true
    },
  },
  c: true,
}
```

**Comment by P4sca1** [comment]
I think the most problematic part and why this is not implemented yet is type-safety.
Getting typings right with string arrays, especially when it contains magic values or relation fields like `*` or `customer.*` can be very challenging.
I personally prefer type-safety over a less verbose syntax.

**Comment by henk23** [comment]
I've been really getting into TypeScript in the last year or so. I've seen many things in the name of so-called type "safety", in a variety of projects, big and small: 

- Code being a lot (!) less readable
- Devs tolerating 50-100% more "code", type definitions and whatnot, that needs to be read, understood and maintained
- Devs being a lot slower as a result, unsurprisingly
- Devs using `as string` a lot (!) to make the red squiggly line in the IDE go away
- Lead devs being super hyped about all these questionable developments

Now I can add another thing: - Cool, helpful features not being implemented in libraries

What I have yet to see is a case of bugs actually being prevented as a result of type-safety. On the contrary, I have seen sooo many bugs as a result of introducing "safe" typing because devs rely on "when it's not red in the IDE, it will not cause problems", abandoning all logic and common sense. For me it's gotten to a point where "type-safety" is so utterly absurd, like it's something out of a Monty Python sketch. But: This is not the place to discuss this.

If this was ever implemented, it could be done like this:

- As an opt-in. Devs who prefer more verbose code, should still be able to do it that way, of course
- With a big red warning in the docs "if you use this feature, you will abandon all type-safety and you will be completely on your own, type-wise, you have been warned!"

What's the worst that can happen? A dev will select an id, forget that it's a number, try to match it against some regex. There will be an error thrown: `Uncaught TypeError: order.id.match is not a function`. The dev remembers they need a string to match. Or SQL throws: `Unknown column firatName in field list` - a typo, it happens. The dev fixes the bug and goes on with their life. Nobody got hurt. ^^

Sorry for the rant. This whole topic has been bothering me for a while. It's not directed at you, @P4sca1, but at the state of the js world in general, you just happened to trigger me to finally type it out.
